File: Lib/cpp/Yolo/Yolo.cpp
#include <fstream>
#include "Yolo.h"

using namespace std;
using namespace cv;
using namespace dnn;

Yolo::Yolo() {}

Yolo::Yolo(const char* cfg, const char* names, const char* weights, GraphicMode mode, WithCUDA cuda, Size size, float confThresh, float nmsThresh)
{

    _confThresh = confThresh;
    _nmsThresh = nmsThresh;
    _blob_size = size;
    _mode = mode;

    string buf;
    ifstream ifs(names);
    while (getline(ifs, buf))
    {
        _labels.push_back(buf);
        auto color = Scalar(rand() % 256, rand() % 256, rand() % 256);
        _colors.push_back(color);
    }

    _net = readNetFromDarknet(cfg, weights);
    if (cuda == On)
    {
        _net.setPreferableBackend(DNN_BACKEND_CUDA);
        _net.setPreferableTarget(DNN_TARGET_CUDA);
    }
    else
    {
        _net.setPreferableBackend(DNN_BACKEND_OPENCV);
        _net.setPreferableTarget(DNN_TARGET_CPU);
    }

}

void Yolo::Run(Mat& img, YoloResults& results)
{

    Process(img, results);

    switch (_mode)
    {
    case None:
        break;
    case Points:
        DrawPoints(img, results);
        break;
    case Boxes:
        DrawRects(img, results);
        break;
    case BoxesWithLabels:
        DrawRectsWithLabels(img, results);
        break;
    }

}

void Yolo::Process(Mat& img, YoloResults& results)
{

    results.count = 0;
    results.classIds.clear();
    results.centers.clear();
    results.confidences.clear();
    results.boxes.clear();

    vector<int> classIds;
    vector<Rect2d> boxes;
    vector<float> confidences;

    vector<Mat> outs;
    vector<String> outNames = _net.getUnconnectedOutLayersNames();
    auto blob = blobFromImage(img, 1.0f / 255, _blob_size, Scalar(), true, false);
    _net.setInput(blob);
    _net.forward(outs, outNames);

    for (Mat pred : outs)
    {
        float* data = (float*)pred.data;
        for (int i = 0; i < pred.rows; i++, data += pred.cols)
        {
            Point classIdPoint;
            double confidence = (double)data[4];
            if (confidence > _confThresh)
            {
                double probability;
                minMaxLoc(pred.row(i).colRange(5, pred.cols), 0, &probability, 0, &classIdPoint);
                if (probability > _confThresh)
                {
                    auto centerX = (int)(data[0] * img.cols);
                    auto centerY = (int)(data[1] * img.rows);
                    auto width = (int)(data[2] * img.cols);
                    auto height = (int)(data[3] * img.rows);
                    classIds.push_back(classIdPoint.x);
                    confidences.push_back(confidence);
                    boxes.push_back(Rect2d(centerX, centerY, width, height));
                }
            }
        }
    }

    MatShape indices;
    NMSBoxes(boxes, confidences, _confThresh, _nmsThresh, indices);

    for (int i : indices)
    {
        auto box = boxes[i];
        results.count++;
        results.classIds.push_back(classIds[i]);
        results.centers.push_back(Point(box.x, box.y));
        results.confidences.push_back(confidences[i]);
        results.boxes.push_back(box);
    }

}

void Yolo::DrawPoints(Mat& img, YoloResults& results)
{

    for (int i = 0; i < results.count; i++)
    {
        circle(img, results.centers[i], 3, _colors[results.classIds[i]], FILLED);
    }

}

void Yolo::DrawRects(Mat& img, YoloResults& results)
{

    for (int i = 0; i < results.count; i++)
    {
        auto box = results.boxes[i];
        int left = box.x - box.width / 2;
        int top = box.y - box.height / 2;
        rectangle(img, Point(left, top), Point(left + box.width, top + box.height), _colors[results.classIds[i]]);
    }

}

void Yolo::DrawRectsWithLabels(Mat& img, YoloResults& results)
{

    for (int i = 0; i < results.count; i++)
    {
        auto box = results.boxes[i];
        int left = box.x - box.width / 2;
        int top = box.y - box.height / 2;
        rectangle(img, Point(left, top), Point(left + box.width, top + box.height), _colors[results.classIds[i]]);
        auto label = _labels[results.classIds[i]] + format(" %d", (int)(results.confidences[i] * 100)) + "%";
        int baseLine;
        auto labelSize = getTextSize(label, FONT_HERSHEY_SIMPLEX, 0.3, 1, &baseLine);
        rectangle(img, Point(left, max(top, labelSize.height) - labelSize.height), Point(left + labelSize.width, max(top, labelSize.height) + baseLine), _colors[results.classIds[i]], FILLED);
        putText(img, label, Point(left, max(top, labelSize.height)), FONT_HERSHEY_SIMPLEX, 0.3, Scalar());
    }

}

File: Lib/cs/Husty.OpenCvSharp/Yolo.cs
using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using OpenCvSharp;
using OpenCvSharp.Dnn;

namespace Husty.OpenCvSharp
{

    public enum DrawingMode { Off, Rectangle, Point }

    /// <summary>
    /// Well-known object detection algorithm
    /// </summary>
    public class Yolo
    {

        // ------- Fields ------- //

        private readonly DrawingMode _draw;
        private readonly Net _net;
        private readonly float _threshold;
        private readonly float _nmsThreshold;
        private readonly string[] _labels;
        private readonly Size _blobSize;
        private readonly Scalar[] _colors = Enumerable.Repeat(false, 80).Select(x => Scalar.RandomColor()).ToArray();


        // ------- Constructor ------- //

        /// <summary>
        /// Initialize detector
        /// </summary>
        /// <param name="cfg">(.cfg) file</param>
        /// <param name="names">(.names) file</param>
        /// <param name="weights">(.weights) file</param>
        /// <param name="blobSize">Must be multiple of 32</param>
        /// <param name="draw">Off or Point or Rectangle</param>
        /// <param name="confThresh">Threshold of 0.0-1.0 confidence value</param>
        /// <param name="nmsThresh">threshold of 0.0-1.0 NMS value</param>
        public Yolo(string cfg, string names, string weights, Size blobSize, DrawingMode draw = DrawingMode.Rectangle, float confThresh = 0.5f, float nmsThresh = 0.3f)
        {
            _blobSize = blobSize;
            _draw = draw;
            _threshold = confThresh;
            _nmsThreshold = nmsThresh;
            _labels = File.ReadAllLines(names).ToArray();
            _net = CvDnn.ReadNetFromDarknet(cfg, weights);
        }


        // ------- Methods ------- //

        /// <summary>
        /// Inference one frame
        /// </summary>
        /// <param name="frame">Input & output image</param>
        /// <param name="results">Holding 'YoloResult' class</param>
        public void Run(ref Mat frame, out YoloResults results)
        {
            var blob = CvDnn.BlobFromImage(frame, 1.0 / 255, _blobSize, new Scalar(), true, false);
            _net.SetInput(blob);
            var outNames = _net.GetUnconnectedOutLayersNames();
            var outs = outNames.Select(_ => new Mat()).ToArray();
            _net.Forward(outs, outNames);
            results = new YoloResults(GetResults(outs, frame));
        }

        private List<(string Label, float Confidence, Point Center, Size Size, Rect Box)> GetResults(Mat[] output, Mat image)
        {
            var classIds = new List<int>();
            var confidences = new List<float>();
            var probabilities = new List<float>();
            var centers = new List<Point>();
            var boxes = new List<Rect>();
            var w = image.Width;
            var h = image.Height;
            unsafe 
            {
                foreach (var pred in output)
                {
                    var p = (float*)pred.Data;
                    for (var i = 0; i < pred.Rows; i++, p += pred.Cols)
                    {
                        var confidence = p[4];
                        if (confidence > _threshold)
                        {
                            Cv2.MinMaxLoc(pred.Row(i).ColRange(5, pred.Cols), out _, out Point classIdPoint);
                            var probability = p[classIdPoint.X + 5];
                            if (probability > _threshold)
                            {
                                var centerX = (int)(p[0] * w);
                                var centerY = (int)(p[1] * h);
                                var width = (int)(p[2] * w);
                                var height = (int)(p[3] * h);
                                classIds.Add(classIdPoint.X);
                                confidences.Add(confidence);
                                probabilities.Add(probability);
                                centers.Add(new Point(centerX, centerY));
                                boxes.Add(new Rect(centerX - width / 2, centerY - height / 2, width, height));
                            }
                        }
                    }
                }
            }
            var results = new List<(string Label, float Confidence, Point Center, Size Size, Rect Rectangle)>();
            CvDnn.NMSBoxes(boxes, confidences, _threshold, _nmsThreshold, out int[] indices);
            foreach (var i in indices)
            {
                results.Add((_labels[classIds[i]], confidences[i], centers[i], new Size(boxes[i].Width, boxes[i].Height), boxes[i]));
                switch (_draw)
                {
                    case DrawingMode.Off:
                        break;
                    case DrawingMode.Point:
                        DrawPoint(image, classIds[i], centers[i]);
                        break;
                    case DrawingMode.Rectangle:
                        DrawRect(image, classIds[i], confidences[i], centers[i], new Size(boxes[i].Width, boxes[i].Height));
                        break;
                }
            }
            return results;
        }

        private void DrawPoint(Mat image, int classes, Point center)
        {
            image.Circle(center.X, center.Y, 3, _colors[classes], 5);
        }

        private void DrawRect(Mat image, int classes, float confidence, Point center, Size size)
        {
            var label = $"{_labels[classes]}{confidence * 100:0}%";
            var x1 = Math.Max(center.X - size.Width / 2, 0);
            var y1 = Math.Max(center.Y - size.Height / 2, 0);
            var x2 = Math.Min(center.X + size.Width / 2, image.Width);
            var y2 = Math.Min(center.Y + size.Height / 2, image.Height);
            Cv2.Rectangle(image, new Rect(x1, y1, x2 - x1, y2 - y1), _colors[classes], 2);
            var textSize = Cv2.GetTextSize(label, HersheyFonts.HersheyTriplex, 0.3, 0, out var baseline);
            Cv2.Rectangle(image, new Rect(new Point(x1, y1 - textSize.Height - baseline),
                new Size(textSize.Width, textSize.Height + baseline)), _colors[classes], Cv2.FILLED);
            var textColor = Cv2.Mean(_colors[classes]).Val0 < 70 ? Scalar.White : Scalar.Black;
            Cv2.PutText(image, label, new Point(x1, y1 - baseline), HersheyFonts.HersheyTriplex, 0.3, textColor);
        }

    }

    /// <summary>
    /// Accumulating detection results
    /// </summary>
    public class YoloResults : IEnumerable<(string Label, float Confidence, Point Center, Size Size, Rect Box)>, IEnumerator<(string Label, float Confidence, Point Center, Size Size, Rect Box)>
    {

        public YoloResults(IEnumerable<(string Label, float Confidence, Point Center, Size Size, Rect Box)> results)
        {
            Labels = new List<string>();
            Confidences = new List<float>();
            Centers = new List<Point>();
            Sizes = new List<Size>();
            Boxes = new List<Rect>();
            foreach (var result in results)
            {
                Labels.Add(result.Label);
                Confidences.Add(result.Confidence);
                Centers.Add(result.Center);
                Sizes.Add(result.Size);
                Boxes.Add(result.Box);
                Count++;
            }
        }


        private int position = -1;

        /// <summary>
        /// Class labels from (.names) file, such as 'person'.
        /// </summary>
        public List<string> Labels { private set; get; }

        /// <summary>
        /// Range of value is 0.0 ~ 1.0
        /// </summary>
        public List<float> Confidences { private set; get; }

        /// <summary>
        /// Detected center points
        /// </summary>
        public List<Point> Centers { private set; get; }

        /// <summary>
        /// Sizes of detected rectangles
        /// </summary>
        public List<Size> Sizes { private set; get; }

        /// <summary>
        /// Detected rectangles
        /// </summary>
        public List<Rect> Boxes { private set; get; }

        /// <summary>
        /// Detected count
        /// </summary>
        public int Count { private set; get; }

        public (string Label, float Confidence, Point Center, Size Size, Rect Rectangle) this[int i]
            => (Labels[i], Confidences[i], Centers[i], Sizes[i], Boxes[i]);

        public IEnumerator<(string Label, float Confidence, Point Center, Size Size, Rect Box)> GetEnumerator() => this;

        IEnumerator IEnumerable.GetEnumerator() => this;

        public bool MoveNext() => ++position < Count;

        public (string Label, float Confidence, Point Center, Size Size, Rect Box) Current
            => (Labels[position], Confidences[position], Centers[position], Sizes[position], Boxes[position]);

        object IEnumerator.Current => Current;

        public void Dispose() { position = -1; }

        public void Reset() { position = -1; }

    }

}
